File: Abstraction.java
// Abstract base class
abstract class Animal {
    abstract void makeSound();  // Abstract method
}

// Derived class: Dog
class Dog extends Animal {
    void makeSound() {
        System.out.println("Dog says: Woof!");
    }
}

// Derived class: Cat
class Cat extends Animal {
    void makeSound() {
        System.out.println("Cat says: Meow!");
    }
}

// Derived class: Cow
class Cow extends Animal {
    void makeSound() {
        System.out.println("Cow says: Moo!");
    }
}

// Main class
public class Main {
    public static void main(String[] args) {
        Animal[] animals = { new Dog(), new Cat(), new Cow() };

        for (Animal a : animals) {
            a.makeSound();  // Polymorphic call
        }
    }
}

File: Abstraction.py
from abc import ABC, abstractmethod

# Abstract Base Class
class Animal(ABC):
    @abstractmethod
    def make_sound(self):
        pass  # Abstract method, no body

# Derived class: Dog
class Dog(Animal):
    def make_sound(self):
        print("Dog says: Woof!")

# Derived class: Cat
class Cat(Animal):
    def make_sound(self):
        print("Cat says: Meow!")

# Derived class: Cow
class Cow(Animal):
    def make_sound(self):
        print("Cow says: Moo!")

# Main function
def main():
    animals = [Dog(), Cat(), Cow()]

    for animal in animals:
        animal.make_sound()  # Polymorphic call

if __name__ == "__main__":
    main()
